### PR TITLE
figurl: implement state sync

### LINKED ISSUE #110: figurl: sync state between figures

### PR DESCRIPTION
**Pull request by magland**
Fixes #110

### CODE REVIEW

**Comment by magland** [comment]
See the following example: https://github.com/magland/sortingview/blob/magland/issue110/devel/figurl_sync.py

It generates two avg wageform views that have selectedUnitIds in sync

image

**Comment by magland** [comment]
Note that this PR is based on magland/issue107 branch, so the other PR should be merged first. Then this one should get rebased on top of the figurl branch prior to merging.

File: src/figurl/useSyncHive.ts [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+// Hive contains all the value/setValue pairs associated with sync IDs.
+// There should be only one Hive instance per browser session.
+// This is instantiated inside the useSyncHive() hook below.
+class Hive {
+    // all the value/setValue pairs associated with sync IDs
+    #items: {[key: string]: {value: any, setValue: (a: any)=> void}} = {}
+    // The constructor takes an onChange() callback
+    constructor(private onChange: () => void) {
+    }
```
Comment (jsoules) [inline src/figurl/useSyncHive.ts:14]: What's the significance of a body-less constructor in JS? Does it automatically store the parameter in an instance variable or something?

File: src/figurl/useSyncHive.ts [hunk 2]
```diff
@@ -0,0 +1,135 @@
+import { JSONStringifyDeterministic } from "kachery-js/types/kacheryTypes"
+import { useCallback, useMemo, useState } from "react"
```
Comment (jsoules) [inline src/figurl/useSyncHive.ts:2]: Just noting that the comments in this class make it super clear to follow what's going on. That's awesome!

**Review by jsoules — COMMENTED** [review]
This looks good to me so far from reading the code. I haven't had the chance to kick the tires on it extensively in practice yet, but I'll look forward to doing so soon. If you'd like to merge in the meantime, feel free, or if you'd like to wait, that's also fine by me.

File: src/figurl/useSyncHive.ts (hunk 1 above)
Comment (magland) [reply src/figurl/useSyncHive.ts:14]: yes, with the private keyword the onChange becomes a private member.

**Pull request merged**